lib/upipe-modules/upipe_sync.c

```diff
@@ -0,0 +1,1083 @@
+        uint64_t cr = 0;
+        uref_clock_get_cr_sys(uref, &cr);
+        upipe_err_va(upipe, "%s() picture too late by %" PRIu64 "ms, drop pic, recept %" PRIu64 "",
+            __func__, (now - pts) / 27000, (now - cr) / 27000);
+        uref_free(uref);
+        return;
+    }
+
+    //upipe_dbg_va(upipe, "push PTS in %" PRIu64 " ms", (pts - now) / 27000);
+
+    /* buffer pic */
+    ulist_add(&upipe_sync->urefs, uref_to_uchain(uref));
+
+
+    /* timer already active */
+    if (upipe_sync->upump)
+        return;
+
+    /* need upump mgr */
+    if (!ubase_check(upipe_sync_check_upump_mgr(upipe_sync_to_upipe(upipe_sync))))
+        return;
+
+    /* start timer */
+    upipe_sync->pts = pts;
+    upipe_sync_wait_upump(upipe_sync_to_upipe(upipe_sync), pts - now, cb);
+}
+
+/** @internal @This allocates a sync pipe.
+ *
+ * @param mgr common management structure
+ * @param uprobe structure used to raise events
+ * @param signature signature of the pipe allocator
+ * @param args optional arguments
+ * @return pointer to upipe or NULL in case of allocation error
+ */
+static struct upipe *upipe_sync_alloc(struct upipe_mgr *mgr,
+                                      struct uprobe *uprobe,
+                                      uint32_t signature, va_list args)
+{
+    struct upipe *upipe = upipe_sync_alloc_void(mgr, uprobe, signature, args);
+    if (unlikely(upipe == NULL))
+        return NULL;
+
+    struct upipe_sync *upipe_sync = upipe_sync_from_upipe(upipe);
+
+    upipe_sync->latency = 0;
+    upipe_sync->pts = 0;
+    upipe_sync->ticks_per_frame = 0;
+    upipe_sync->frame_idx = 0;
+    upipe_sync->uref = NULL;
+    ulist_init(&upipe_sync->urefs);
+
+    upipe_sync_init_urefcount(upipe);
+    upipe_sync_init_uclock(upipe);
+    upipe_sync_init_upump(upipe);
+    upipe_sync_init_upump_mgr(upipe);
+    upipe_sync_init_output(upipe);
+    upipe_sync_init_sub_subs(upipe);
+    upipe_sync_init_sub_mgr(upipe);
+
+    upipe_throw_ready(upipe);
+    return upipe;
+}
+
+/** @internal @This sets the input flow definition.
+ *
+ * @param upipe description structure of the pipe
+ * @param flow_def flow definition packet
+ * @return an error code
+ */
+static int upipe_sync_set_flow_def(struct upipe *upipe, struct uref *flow_def)
+{
+    struct upipe_sync *upipe_sync = upipe_sync_from_upipe(upipe);
+
+    if (flow_def == NULL)
+        return UBASE_ERR_INVALID;
+
+    const char *def;
+    UBASE_RETURN(uref_flow_get_def(flow_def, &def))
+
+    if (ubase_ncmp(def, "pic.")) {
+        upipe_err_va(upipe, "Unknown def %s", def);
+        return UBASE_ERR_INVALID;
+    }
+
+    UBASE_RETURN(uref_pic_flow_get_fps(flow_def, &upipe_sync->fps));
+
+    uint64_t latency;
+    if (!ubase_check(uref_clock_get_latency(flow_def, &latency)))
+        latency = 0;
+
+    flow_def = uref_dup(flow_def);
+    if (!flow_def)
+        return UBASE_ERR_ALLOC;
+
+    // FIXME : estimated latency added by processing
+    latency += UCLOCK_FREQ / 25;
```

Comment:
Shouldn't this be based on fps ?

Reply:
Yes

Reply:
Do you intend to fix it ?

Reply:
Of course but will need some time to test it properly

Reply:
Isn't it just a matter of adding * fps.den / fps.num ?

Reply:
Yes but I remember having problems with this so want to test for a while

Reply:
Do you want to push it as is ?

Reply:
Yes I guess it doesn't harm

Reply:
Can you squash it ?